src/r1cs/constraint_system.rs

```diff
@@ -104,9 +104,12 @@ pub trait RandomizableConstraintSystem: ConstraintSystem {
     ///     // ...
     /// })
     /// ```
-    fn specify_randomized_constraints<F>(&mut self, callback: F) -> Result<(), R1CSError>
+    fn specify_randomized_constraints<'a: 'constraints, F>(
+        &mut self,
+        callback: F,
+    ) -> Result<(), R1CSError>
```

Comment:
I'm actually unsure whether the additional `'a` lifetime is needed here. IIRC, any closure `FnOnce + 'constraints` should also implement `FnOnce + 'a`, since `'a: 'constraints`, right? I've never had to do a lot of lifetime stuff, but I think the announced experiment of @oleganza should point out anything strange here.

Reply:
first thing i noticed is that now every gadget has to specify additional lifetime parameters

<img width="577" alt="image" src="https://user-images.githubusercontent.com/698/82198700-f5214400-98fc-11ea-9af5-17487ce3123e.png">

Reply:
I'm not sure whether it's possible to mitigate tbqh, but I'm having a look into this. Quite the opportunity to learn a ton about lifetimes for me!

Reply:
I'm actually quite confident now that it's not possible without adding a lifetime parameter to the `RandomizableConstraintSystem`: for a given constraint system `let p = Prover::new(...)`, we need to declare the lifetime of the `FnOnce`s that are stored in the system. Prover can only implement `RandomizableConstraintSystem` when the randomized constraint you are adding outlives that lifetime of the constraintsystem.

Think of it like "Prover only implements RandomizableConstraintSystem for constraints that reference things that outlive the Prover."

... it is indeed quite unfortunate that this has to break the API though, for gadgets that use the RCS.

Reply:
How about upgrading from Fn to FnOnce in this PR, but keeping the 'static requirement, and then separately explore if moving to generic lifetime bound is worth it? Right now i have to propagate 'a in a ton of places in ZkVM, because where I had a `vm::Verifier`, now I have `vm::Verified<'a>`, and it leaks to `PrecomputedTx` that wraps it, etc. I could probably slap 'static at some point inside ZkVM/Spacesuit, but that push the burden of justification on my codebase and I'm just... ugh... not ready for it right now :-P